src/main/java/com/google/devtools/build/lib/runtime/commands/BuildCommand.java

```diff
@@ -64,12 +69,29 @@
   @Override
   public BlazeCommandResult exec(CommandEnvironment env, OptionsParsingResult options) {
     BlazeRuntime runtime = env.getRuntime();
-    List<String> targets;
-    try (SilentCloseable closeable = Profiler.instance().profile("ProjectFileSupport.getTargets")) {
-      // only takes {@code options} to get options.getResidue()
-      targets = ProjectFileSupport.getTargets(runtime.getProjectFileProvider(), options);
-    }
-    if (targets.isEmpty()) {
+    List<String> targets = options.getResidue();
+    BuildRequestOptions buildRequestOptions = options.getOptions(BuildRequestOptions.class);
+    if (!targets.isEmpty()) {
+      if (!buildRequestOptions.buildFile.isEmpty()) {
+        env.getReporter()
+            .handle(Event.error("Command-line target pattern and --build_file cannot both be specified"));
+        return BlazeCommandResult.exitCode(ExitCode.COMMAND_LINE_ERROR);
+      }
+      try (SilentCloseable closeable = Profiler.instance().profile("ProjectFileSupport.getTargets")) {
+        // only takes {@code options} to get options.getResidue()
+        targets = ProjectFileSupport.getTargets(runtime.getProjectFileProvider(), options);
+      }
+    } else if (!buildRequestOptions.buildFile.isEmpty()) {
+      // Works for absolute or relative query file.
+      Path residuePath = env.getWorkingDirectory().getRelative(buildRequestOptions.buildFile);
+      try {
+        targets = Lists.newArrayList(FileSystemUtils.readLines(residuePath, StandardCharsets.UTF_8));
```

Comment:
We should probably check if not targets were specified in the files. If we're not doing this in --query_file it was probably an oversight...

Reply:
The build command explicitly allows calling build without a resisdue/target pattern.

Reply:
I think this was mostly resolved by later commits - IIRC my concern was the target file could be empty and we'd be skip the "Nothing will be built" warning.